### PR TITLE
Add SWIFT_STDLIB_HAS_LOCALE CMake to build stdlib without locale.h and locale based float parsing

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: stdlib/public/stubs/Stubs.cpp [hunk 1]
```diff
@@ -25,16 +25,18 @@
 // Avoid defining macro max(), min() which conflict with std::max(), std::min()
 #define NOMINMAX
 #include <windows.h>
-#else
-#if !defined(__HAIKU__) && !defined(__wasi__)
+#else // defined(_WIN32)
+#if !defined(__HAIKU__) && !defined(__wasi__) && __has_include(<sys/errno.h>)
```
Comment (compnerd) [inline stdlib/public/stubs/Stubs.cpp:29]: I think that the guards here are unnecessary, `__has_inlcude(<sys/errno.h>)` covers everything.  Is there something Im missing?

File: stdlib/public/stubs/Stubs.cpp [hunk 2]
```diff
@@ -44,6 +46,9 @@
 #if defined(__CYGWIN__) || defined(__HAIKU__)
 #include <sstream>
 #endif
+
+#if SWIFT_STDLIB_HAS_LOCALE
+#include <clocale>
 #if defined(__OpenBSD__) || defined(__ANDROID__) || defined(__linux__) || defined(__wasi__) || defined(_WIN32)
```
Comment (compnerd) [inline stdlib/public/stubs/Stubs.cpp:52]: I don't understand the point of this check.  The PR changes the check to the explicit option, so we shouldn't need that anymore.

File: stdlib/public/stubs/Stubs.cpp [hunk 3]
```diff
@@ -44,6 +46,9 @@
...
 #include <locale.h>
```
Comment (compnerd) [inline stdlib/public/stubs/Stubs.cpp:53]: Why do we need this header?  If we are including `clocale` we should have all the same declarations already.

File: stdlib/public/stubs/Stubs.cpp [hunk 4]
```diff
@@ -138,14 +145,14 @@ uint64_t swift_uint64ToString(char *Buffer, intptr_t BufferLength,
                             /*Negative=*/false);
 }
 
-#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__ANDROID__)
+#if !SWIFT_STDLIB_HAS_LOCALE
```
Comment (compnerd) [inline stdlib/public/stubs/Stubs.cpp:148]: I wonder if its nicer to do:

```
#if SWIFT_STDLIB_HAS_LOCALE
#if defined(_WIN32)
...
#else
...
#endif
#endif
```

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Review by mikeash — APPROVED** [review]

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test macOS platform

**Review by compnerd — APPROVED** [review]
Thanks for following up on the cleanups!

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Pull request merged**